packages/pluggableWidgets/datagrid-web/src/utils/settings.ts

```diff
@@ -0,0 +1,116 @@
+    hiddenColumns: Array<IdType<object>>,
+    setHiddenColumns: Dispatch<SetStateAction<Array<IdType<object>>>>,
+    sortBy: Array<SortingRule<object>>,
+    setSortBy: Dispatch<SetStateAction<Array<SortingRule<object>>>>,
+    filters: Filters<object>,
+    setFilters: Dispatch<SetStateAction<Filters<object>>>,
+    widths: ColumnWidth,
+    setWidths: Dispatch<SetStateAction<ColumnWidth>>
+): void {
+    const [loaded, setLoaded] = useState(false);
+
+    const filteredColumns = useMemo(
+        () =>
+            columns.map((c, index) => ({
+                header: typeof c.header === "object" ? c.header.value : c.header,
+                id: index.toString(),
+                hidable: c.hidable
+            })) as Array<{ header: string; id: string; hidable: string }>,
+        [columns]
+    );
+
+    useEffect(() => {
+        if (settings && settings.status !== ValueStatus.Loading && settings.value && !loaded) {
+            const newSettings = JSON.parse(settings.value) as PersistedSettings[];
+            const columns = newSettings.map(columnSettings => ({
+                ...columnSettings,
+                columnId: filteredColumns.find(c => c.header === columnSettings.column)?.id || ""
+            }));
+
+            const extractedSettings = {
+                columnOrder: columns.sort((a, b) => a.order - b.order).map(s => s.columnId),
+                hiddenColumns: columns.filter(s => s.hidden).map(s => s.columnId),
+                sortBy: columns
+                    .filter(s => s.sort)
+                    .map(s => ({
+                        id: s.columnId,
+                        desc: s.sortMethod === "desc"
+                    })),
+                filters: columns.filter(s => s.filter).map(s => ({ id: s.columnId, value: s.filter })),
+                widths: Object.fromEntries(columns.map(s => [s.columnId, s.width]))
+            };
+            setColumnOrder(extractedSettings.columnOrder);
+            setHiddenColumns(extractedSettings.hiddenColumns);
+            setSortBy(extractedSettings.sortBy);
+            setFilters(extractedSettings.filters);
+            setWidths(extractedSettings.widths);
+            setLoaded(true);
+        }
+    }, [settings, loaded, filteredColumns]);
+
+    useEffect(() => {
+        if (settings && settings.status === ValueStatus.Available) {
+            const newSettings = JSON.stringify(
+                createSettings(
+                    {
+                        columnOrder,
+                        hiddenColumns,
+                        sortBy,
+                        filters,
+                        widths
+                    },
+                    filteredColumns
+                ) ?? []
+            );
+            if (!deepEqual(settings.value, newSettings, { strict: true })) {
+                settings.setValue(newSettings);
+            }
+        }
+    }, [columnOrder, hiddenColumns, sortBy, filters, widths]);
```

Comment:
I think `settings` is a dep

Reply:
Nice catch!

Reply:
We need to introduce the eslint rule to prevent this :)

Reply:
I will create a story for that.